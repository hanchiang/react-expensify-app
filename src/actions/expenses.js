// Expenses action creators
import uuid from 'uuid';
import database from '../firebase/firebase';

// ADD_EXPENSE
const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

// Add to database
// Async thunk to add expense to firebase and trigger ADD_EXPENSE action
const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const { description = '', note = '', amount = 0, createdAt = 0 } = expenseData;
    const expense = { description, note, amount, createdAt };

    return database.ref(`users/${uid}/expenses`).push(expense)
      .then((ref) => {
        dispatch(addExpense({
          id: ref.key,
          ...expense
        }));
      });
  }
}

// REMOVE_EXPENSE
const removeExpense = (id = '') => ({
  type: 'REMOVE_EXPENSE',
  id
});

// Remove from database
// Async thunk to remove expense from firebase and trigger REMOVE_EXPENSE
const startRemoveExpense = (id = '') => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses/${id}`).remove()
      .then(() => dispatch(removeExpense(id)));
  }
}

// EDIT_EXPENSE
const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

// Modify data in the database
// Asuync thunk function to modify data in firebase
const startEditExpense = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses/${id}`).update(updates)
      .then(() => dispatch(editExpense(id, updates)));
  }
};

// SET_EXPENSES
const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

// Fetch from database
// Async thunk function to fetch expenses from firebase and trigger SET_EXPENSES action
const startSetExpenses = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const expenses = [];
    return database.ref(`users/${uid}/expenses`)
      .once('value')
      .then((snapshot) => {
        snapshot.forEach((childSnapshot) => {
          expenses.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          })
        });
        dispatch(setExpenses(expenses));
      });
  }
}


export { addExpense, startAddExpense, removeExpense, editExpense, startEditExpense, 
  setExpenses, startSetExpenses, startRemoveExpense };

