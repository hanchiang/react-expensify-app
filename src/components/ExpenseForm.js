// A form to edit or add expense
// Use local state to track form inputs
// Form data to be passed back to the parent component which handles the interaction with redux

// Will be used for AddExpensePage and EditExpensePage
// For editting, the expense's information have to be grabbed from the redux store in EditExpensePage,
// and passed to ExpenseForm as props

import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

// const now = moment();
// console.log(now.format('MMM Do YYYY'));

class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);

        const expense = props.expense;
        this.state = {
            description: expense ? expense.description : '',
            note: expense ? expense.note : '',
            amount: expense ? (expense.amount / 100).toString() : '',
            createdAt: expense ? moment(expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        };

        this.onDescriptionChange = this.onDescriptionChange.bind(this);
        this.onNoteChange = this.onNoteChange.bind(this);
        this.onAmountChange = this.onAmountChange.bind(this);
        this.onDateChange = this.onDateChange.bind(this);
        this.onFocusChange = this.onFocusChange.bind(this);
        this.onSubmit = this.onSubmit.bind(this);
    }

    onDescriptionChange(event) {
        this.setState({ description: event.target.value });
    }

    onNoteChange(event) {
        this.setState({ note: event.target.value });
    }

    onAmountChange(event) {
        const amount = event.target.value;
        // validate amount
        if (!amount || amount.match(/^\d+(\.\d{0,2})?$/)) {
            this.setState({ amount });
        }
    }

    onDateChange(createdAt) {
        if (createdAt) {
            this.setState({ createdAt });
        }
    }

    onFocusChange({ focused }) {
        this.setState({ calendarFocused: focused });
    }

    onSubmit(event) {
        event.preventDefault();
        let error = '';

        if (!this.state.description || !this.state.amount) {
            error = 'Please provide a description and amount';
            this.setState({ error });
        } else {
            this.setState({ error });
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    }

    render() {
        return (
            <form className="form" onSubmit={this.onSubmit}>
                {this.state.error && <p className="form__error">{this.state.error}</p>}
                <input
                    type="text" placeholder="Description" className="text-input"
                    value={this.state.description} onChange={this.onDescriptionChange} autoFocus
                />
                <input
                    type="number" placeholder="Amount" className="text-input"
                    value={this.state.amount} onChange={this.onAmountChange}
                />
                <SingleDatePicker
                    date={this.state.createdAt} // momentPropTypes.momentObj or null
                    onDateChange={this.onDateChange} // PropTypes.func.isRequired
                    focused={this.state.calendarFocused} // PropTypes.bool
                    onFocusChange={this.onFocusChange} // PropTypes.func.isRequired
                    numberOfMonths={1}
                    isOutsideRange={(day) => false}
                />
                <textarea placeholder="Add a note for your expense (optional)"
                    value={this.state.note} onChange={this.onNoteChange} className="textarea">
                </textarea>
                <div>
                    <button className="button">Save expense</button>
                </div>
                
            </form>
        );
    }
}

export default ExpenseForm;