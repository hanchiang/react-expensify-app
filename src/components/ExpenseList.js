import React from 'react';
import { connect } from 'react-redux';
import selectExpense from '../selectors/expenses';

import ExpenseListItem from './ExpenseListItem';

export function ExpenseList(props) {
    return (
        <div>
            {
                props.expenses.length === 0 ? (
                    <p>No expenses</p>
                ) : (
                    props.expenses.map((expense) =>
                        <ExpenseListItem expense={expense} key={expense.id} />
                    )
                )

            }
        </div>
    );
}

const mapStateToProps = (state) => ({ 
    expenses: selectExpense(state.expenses, state.filters)
});

export default connect(mapStateToProps)(ExpenseList);