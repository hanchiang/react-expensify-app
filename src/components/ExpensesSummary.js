import React from 'react';
import numeral from 'numeral';
import { connect } from 'react-redux';

import getVisibleExpenses from '../selectors/expenses';
import getExpensesTotal from '../selectors/expenses-total';

export function ExpensesSummary(props) {
    const { expenseCount, expensesTotal } = props;
    const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
    const formattedExpensesTotal = numeral(expensesTotal / 100).format('$0,0.00');

    return (
        <div>
            {
                expenseCount === 0 ?
                <h2>No expenses</h2> : 
                <h2>Viewing {expenseCount} {expenseWord} totalling {formattedExpensesTotal}</h2>
            }
        </div>
        
    );
}

const mapStateToProps = (state) => {
    const visibleExpenses = getVisibleExpenses(state.expenses, state.filters);
    return {
        expenseCount: visibleExpenses.length,
        expensesTotal: getExpensesTotal(visibleExpenses)
    }
}

export default connect(mapStateToProps)(ExpensesSummary);