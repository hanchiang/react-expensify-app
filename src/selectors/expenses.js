// Expenses selector - Get visible expenses
// Returns filtered and sorted expenses
import moment from 'moment';

// startDate and endDate are moment objects
const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        const startDateMatch = startDate ? startDate.isSameOrBefore(expense.createdAt) : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(expense.createdAt) : true;
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if (sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1;
        } else if (sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1;
        }
    });
};

export default getVisibleExpenses;