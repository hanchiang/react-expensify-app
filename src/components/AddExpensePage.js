import React from 'react';
import { connect } from 'react-redux';

import { startAddExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export function AddExpensePage(props) {
    const handleSubmit = (expense) => {
        props.startAddExpense(expense);
        props.history.push('/');
    }
    return (
        <div>
            <div className="page-header">
                <div className="content-container">
                    <h1 className="page-header__title">Add expense</h1>
                </div>
            </div>
            <div className="content-container">
                <ExpenseForm onSubmit={handleSubmit} />
            </div>
            
            
        </div>
    );
}

const mapDispatchToProps = (dispatch, props) => ({
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(null, mapDispatchToProps)(AddExpensePage);