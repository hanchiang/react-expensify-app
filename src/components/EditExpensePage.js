import React from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

// Use the id from match.params to find the expense object

export function EditExpensePage(props) {
    const handleRemove = (event) => {
        props.startRemoveExpense(props.expense.id);
        props.history.push('/');
    }
    const handleSubmit = (expense) => {
        props.startEditExpense(props.expense.id, expense);
        props.history.push('/');
    }

    return (
        <div>
            <ExpenseForm onSubmit={handleSubmit} expense={props.expense} />
            <button onClick={handleRemove}>Remove</button>
        </div>
    );
}

const mapDispatchToProps = (dispatch, props) => ({
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: (id) => dispatch(startRemoveExpense(id))
    
});

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find(expense => expense.id === props.match.params.id)
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);