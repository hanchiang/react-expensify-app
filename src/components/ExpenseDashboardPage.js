import React from 'react';

import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';

function ExpenseDashboardPage() {
    return (
        <div>
            <ExpenseListFilters />
            <ExpenseList />
        </div>
    )
}

export default ExpenseDashboardPage;