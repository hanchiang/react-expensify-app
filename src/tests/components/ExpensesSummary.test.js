import React from 'react';
import { shallow } from 'enzyme';

import expenses from '../fixtures/expenses';
import { altFilters3, altFilters4 } from '../fixtures/filters';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import getVisibleExpenses from '../../selectors/expenses';
import getExpensesTotal from '../../selectors/expenses-total';

test('should show expenses summary for 1 expense', () => {
    const finalExpenses = getVisibleExpenses(expenses, altFilters3);
    const expenseCount = finalExpenses.length;
    const expensesTotal = getExpensesTotal(finalExpenses);
    const wrapper = shallow(<ExpensesSummary expenseCount={expenseCount} expensesTotal={expensesTotal} />);
    expect(wrapper).toMatchSnapshot();
});

test('should show expenses summary for 2 expenses', () => {
    const finalExpenses = getVisibleExpenses(expenses, altFilters4);
    const expenseCount = finalExpenses.length;
    const expensesTotal = getExpensesTotal(finalExpenses);
    const wrapper = shallow(<ExpensesSummary expenseCount={expenseCount} expensesTotal={expensesTotal} />);
    expect(wrapper).toMatchSnapshot();
});