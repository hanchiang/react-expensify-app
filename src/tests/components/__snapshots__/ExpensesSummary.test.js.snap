// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should show expenses summary for 1 expense 1`] = `
<div>
  <h2>
    Viewing 
    1
     
    expense
     totalling 
    $1,305.02
  </h2>
</div>
`;

exports[`should show expenses summary for 2 expenses 1`] = `
<div>
  <h2>
    Viewing 
    2
     
    expenses
     totalling 
    $1,658.06
  </h2>
</div>
`;
